test/integration/sds_dynamic_integration_test.cc

```diff
@@ -974,5 +978,168 @@ TEST_P(SdsCdsIntegrationTest, BasicSuccess) {
   test_server_->waitForGaugeEq("cluster_manager.active_clusters", 3);
 }
 
+class SdsDynamicDownstreamPrivateKeyIntegrationTest : public SdsDynamicDownstreamIntegrationTest {
+public:
+  envoy::extensions::transport_sockets::tls::v3::Secret getCurrentServerPrivateKeyProviderSecret() {
+    envoy::extensions::transport_sockets::tls::v3::Secret secret;
+    ProtobufWkt::Struct val;
+    (*val.mutable_fields())["private_key_file"].set_string_value(
+        TestEnvironment::temporaryPath("root/current/serverkey.pem"));
+    (*val.mutable_fields())["expected_operation"].set_string_value("sign");
+    (*val.mutable_fields())["sync_mode"].set_bool_value(true);
+
+    // check the key type and set the mode accordingly
+    std::string mode;
+    std::string private_key = TestEnvironment::readFileToStringForTest(
+        TestEnvironment::temporaryPath("root/current/serverkey.pem"));
+    bssl::UniquePtr<BIO> bio(
+        BIO_new_mem_buf(const_cast<char*>(private_key.data()), private_key.size()));
+    bssl::UniquePtr<EVP_PKEY> pkey(PEM_read_bio_PrivateKey(bio.get(), nullptr, nullptr, nullptr));
+
+    EXPECT_TRUE(pkey != nullptr);
+
+    if (EVP_PKEY_id(pkey.get()) == EVP_PKEY_RSA) {
+      mode = "rsa";
+    } else if (EVP_PKEY_id(pkey.get()) == EVP_PKEY_EC) {
+      mode = "ecdsa";
+    }
+    (*val.mutable_fields())["mode"].set_string_value(mode);
+
+    secret.set_name(server_cert_rsa_);
+    auto* tls_certificate = secret.mutable_tls_certificate();
+    tls_certificate->mutable_certificate_chain()->set_filename(
```

Comment:
Similarly, prefer to build from YAML if possible.

Reply:
Fixed.